### DISCUSSION TITLE
Adafruit nrf52480 Feather: Can only send 4 byte MIDI SysEx messages

### DISCUSSION THREAD

**Discussion by boomninjavanish** [FortySevenEffects/arduino_midi_library#316] [Q&A]
Hi!

I have been working on a project where I am required to send several floating point numbers via MIDI over BLE. I decided to use real-time sysex messages as no other message within in the MIDI v1.0 spec would contain the 7 plus floating point numbers I need to receive from the feather.

However, I am finding that when sending them as raw bytes, I am only receiving a four byte message consisting of the sysex start byte (0xF0), two bytes of the data, then the sysex end byte (0xF7). The rest of the data is then interpreted as whatever either the MIDI.h or BLEMidi.h libraries interpret the rest of the bytes of data to be.

I have tried to use both versions of the MIDI.sendSysEx() function but whether I provide the start and end bytes or not, the message data is still reduced to two bytes with the start/end bytes on either side. Digging through the source code of both FourtySevenEffects MIDI and the Adafruit BLE MIDI libraries has me wondering if there is something I am not setting up properly in MIDI.h or a problem with how the BLEMidi.h send functions detect sysex messages and tries to terminate them properly. Either way, my eyeballs are bleeding and I am frustrated.

For simplicity, I reduced my code and placed it in the blemidi example for testing. I removed the scale playback and inserted this:

```c++
union DataArray {
    float values[2];
    uint8_t bytes[8];
  } data;

  data.bytes[0] = 0x7f; // id = real-time message
  data.bytes[1] = 0x00; // dev id
  data.bytes[2] = 0x01; // sub id #1
  data.bytes[3] = 0x02; // sub id #2

  data.values[1] = 22.7;

  uint8_t test[10] = {0xf0, 0x7f, 0x00, 0x01, 0x02, 0xff, 0xff, 0xff, 0xff, 0xf7};

  //MIDI.sendSysEx(10, data.bytes);
  MIDI.sendSysEx(10, test, true); 
```

Note that I tried sending my unionized memory buffer and the `test[10]` buffer to no avail.

Here is the output from Snoize's MIDI Monitor:

```
21:16:49.944	From Feather nRF52840 Express Bluetooth	SysEx		Universal Real Time 4 bytes	F0 7F 00 F7
21:16:50.002	From Feather nRF52840 Express Bluetooth	Reset		
21:16:50.002	From Feather nRF52840 Express Bluetooth	Reset		
21:16:50.004	From Feather nRF52840 Express Bluetooth	Reset		
... [repeats indefinitely]
```
Thanks in advance for the help! Even if you can point me in the right direction or tell me how to just bit bang the stuff over, I will super grateful!

**Comment by franky47** [comment]
This could be an issue with chunking of the SysEx into BLE packets, maybe @lathoub can chime in as he worked on that transport.

Reset corresponds to the 0xff messages, so it's possible that:
1. The first four bytes of data are sent as SysEx, with the 0xf0 start
2. The next bytes of data are incorrectly sent as RealTime Reset messages (0xff)

If we can manage to fix sending a hardcoded SysEx buffer, then we can move on to using unions to encode the data.

**Comment by lathoub** [comment]
> However, I am finding that when sending them as raw bytes, I am only receiving a four byte message consisting of the sysex start byte (0xF0), two bytes of the data, then the sysex end byte (0xF7). The rest of the data is then interpreted as whatever either the MIDI.h or BLEMidi.h libraries interpret the rest of the bytes of data to be.

You are **sending** 10 bytes using the BLE Transport of this lib. The underlying code simply send out the buffer 'as is' (there is no interpreting in sending the buffer)

https://github.com/lathoub/Arduino-BLE-MIDI/blob/929c2fc04962672ddba903a618601bea44ee1f3e/src/BLEMIDI_Transport.h#L105

Are you also using this lib for receiving the BLE bytes? (or is Snoize's the receiving app?).

**Reply by boomninjavanish** [reply]
> You are sending 10 bytes using the BLE Transport of this lib. The underlying code simply send out the buffer 'as is' (there is no interpreting in sending the buffer)

Correct. It doesn't matter which set of bytes or the length of the array being sent. The other side receives the first two bytes of data surrounded by the SysEx start and end bytes. The next messages are the rest of the bytes from the data array.

> Are you also using this lib for receiving the BLE bytes? (or is Snoize's the receiving app?).

I am sending SysEx messages using the FortySevenEffects MIDI library using BLE as the transport layer. For testing, I was watching the midi data using [this app](https://www.snoize.com/midimonitor/) for macOS. It is a MIDI monitor developed by Snoize. 

Thanks for looking into this. I am not concerned about the encoding/decoding of the floating point data--I developed that already. I just need help getting the sending of SysEx messages working. Thanks again!

**Reply by boomninjavanish** [reply]
@lathoub 

I am now wondering if it may be an Adafruit problem. I am using their bluefruit.h lib which includes the following code:

https://github.com/adafruit/Adafruit_nRF52_Arduino/blob/8a3978666c51d5a940b038f4c45322aa4cdbbc2f/libraries/Bluefruit52Lib/src/services/BLEMidi.cpp#L248

Am I even using your BLE transport at this point? Forgive my ignorance.

**Reply by lathoub** [reply]
> Am I even using your BLE transport at this point? Forgive my ignorance.

No

**Reply by boomninjavanish** [reply]
One last question then I will stop bothering you, @lathoub. Without switching to another board, is there a way to utilize your transport instead of theirs? I really do appreciate the help.

**Reply by lathoub** [reply]
I do not believe so. I had started the [development to support the nrf52480](https://github.com/lathoub/Arduino-BLE-MIDI/blob/master/src/hardware/BLEMIDI_nRF52.h), but my hardware failed  and have not continued development. Maybe you can?
